`--announce_rc` includes flags set by policy

illicitonion opened the pull request:
Users often use `--announce_rc` to debug why some specific flag value is or isn't being applied. Before this commit, flags set by `--invocation_policy` were silently applied without announcing them through this interface, leading to hard to diagnose configuration issues.

illicitonion committed:
`--announce_rc` includes flags set by policy

Users often use `--announce_rc` to debug why some specific flag value is
or isn't being applied. Before this commit, flags set by
`--invocation_policy` were silently applied without announcing them
through this interface, leading to hard to diagnose configuration
issues.

illicitonion commented:
/cc @michajlo

illicitonion commented:
@michajlo please could I nudge you at this PR (and less urgently https://github.com/bazelbuild/bazel/pull/17042)? Thanks!

michajlo commented:
I'm OOO, can try to take a look when I'm back ~next week.

michajlo commented on src/main/java/com/google/devtools/build/lib/runtime/BlazeOptionHandler.java:
```diff
@@ -361,6 +366,80 @@ DetailedExitCode parseOptions(List<String> args, ExtendedEventHandler eventHandl
+
+    List<String> commandNamesToParse = getCommandNamesToParse(commandAnnotation);
+
+    for (InvocationPolicyOuterClass.FlagPolicy flagPolicy :
```
having a hard time following this loop. instead of effectively duplicating the invocation-policy enforcement code, could we tie the reporting into InvocationPolicyEnforcer#enforce?

illicitonion replied:
Done - this required a bit of restructuring to be able to preserve the command the flag applied to, but it's hopefully all wired up ok.

michajlo commented on src/test/java/com/google/devtools/build/lib/runtime/BlazeOptionHandlerTest.java:
```diff
@@ -1084,6 +1092,93 @@ public void testParseOptions_2LongChains() {
 
+  @Test
+  public void testParseOptions_invocationPolicy() throws Exception {
+    setUp(makeInvocationPolicy());
```
is it possible to inline some of the details of our made-up invocation policy in the test method? this will make it more clear what we have coverage for, as well as provide more immediate context for the assertions. we can still have factory helpers for the boilerplate to improve readability. for example...

```
  setUp(
      InvocationPolicy.newBuilder()
          .addFlagPolicy(setValue("parent", "parent_test_string", "parent-for-parent", ALLOW_OVERRIDES))
         .addFlagPolicy(setValue(...))
         ...
```

illicitonion replied:
Inlined, and added comments to each FlagPolicy explaining why it's there and what's special about it.

michajlo reviewed (commented):
we'll want to update the announce_rc documentation for it's expanded scope (i think this expansion of scope is tolerable)

is there a shell test we could add a smoke test to? just to demonstrate end-to-end what we expect output to look like, and that it works.

illicitonion committed:
Review comments

illicitonion committed:
Add UI test

illicitonion committed:
Update site docs for announce_rc

illicitonion commented:
> we'll want to update the announce_rc documentation for it's expanded scope (i think this expansion of scope is tolerable)

Updated site/en/docs/user-manual.md - let me know if there's anywhere else you think we should update.

> is there a shell test we could add a smoke test to? just to demonstrate end-to-end what we expect output to look like, and that it works.

Added a small shell test.

michajlo commented:
Alas, you seem to have caught me right before another big chunk of leave. Pretty swamped preparing for that at the moment, but I'll try to take a look if I get some cycles, and if not pass it off to a teammate.

meisterT commented:
@illicitonion in general seems sensible - do you want to rebase?
